zip-0239.rst

```diff
@@ -0,0 +1,175 @@
+txid after rejecting a previous transaction with that txid — would allow a third
+party to interfere with transaction relay by malleating a transaction's witness and
+announcing the resulting invalid transaction to nodes, preventing relay of the valid
+version of the transaction as well.
+
+This inefficiency was present in Bitcoin for almost 3 years after activation of its
+Segwit upgrade [#bip-0141]_, until the adoption of BIP 339 [#bip-0339]_. The latter
+BIP specifies a way to use the Segwit "wtxid" (which commits to both effecting and
+witness data) in place of the txid when announcing and fetching transactions.
+In Zcash, the analogous identifier is also called the wtxid, but it encodes the pair
+(txid, auth_digest).
+
+This ZIP is modelled after BIP 339: it adds a ``MSG_WTX`` inv type to the peer-to-peer
+protocol, analogous to BIP 339's ``MSG_WTX`` inv type, that announces transactions by
+their wtxid. Note that the encoding and length of a Zcash wtxid is different to that
+of a Bitcoin wtxid.
+
+This ZIP does not introduce any equivalent of BIP 339's ``wtxidrelay`` message,
+since that is not needed for compatibility given that Zcash full nodes are required to
+support ``MSG_WTX`` based on the negotiated peer protocol version (see `Deployment`_).
+
+
+Specification
+=============
+
+A new inv type ``MSG_WTX`` (0x00000005) is added, for use in both ``inv`` messages
+and ``getdata`` requests, indicating that the hash being referenced is the wtxid
+(i.e. the 64-byte value ``txid`` || ``auth_digest``). This inv type MUST be used
+when announcing v5 transactions. The ``txid`` and ``auth_digest`` are as defined in
+[#zip-0244]_.
+
+In the case of ``getdata`` requests, the format of a v5 transaction obtained by a
+``MSG_WTX`` inv type is as given in the Zcash protocol specification.
+[#protocol-txnencodingandconsensus]_
+
+An ``inv`` or ``getdata`` message MUST NOT use the ``MSG_WTX`` inv type for v4
+or earlier transactions, or on peer connections that have not negotiated at least
+the peer protocol version specified in `Deployment`_.
+
+Note that ``MSG_WTX`` might also be used for future transaction versions after v5.
+Since such versions are not currently consensus-valid, this is left unspecified
+for now.
+
+``MSG_TX`` and ``MSG_WTX`` entries may be mixed in arbitrary order in an ``inv``
+message or a ``getdata`` message. Since these entry types are of different lengths
+(36 bytes or 68 bytes respectively including the 4-byte type field), this implies
+that the size of the message is not determined by its initial ``count`` field, and
+has to be determined by parsing the whole message.
+
+**Open issue:**
+
+* Should we redefine ``inv`` and ``getdata`` to segregate ``MSG_TX``
+  and ``MSG_WTX``, making these messages simpler / more efficient to parse?
+  This could be done by adding new ``invw`` and ``getdataw`` messages.
+  See [#p2p-inv]_ and [#p2p-getdata]_ for how ``inv`` and ``getdata``
+  respectively are currently defined in Bitcoin.
```

Comment:
Seems unnecessary at this stage?

Reply:
Well, the argument from simplicity would be moot if we had to support both variants. Either we make this change now, or we don't make it, I think.

Reply:
The current implementations of ["inv" message parsing](https://github.com/zcash/zcash/blob/8c95ceba46aa204f923f6c7d69a6bab912db9758/src/main.cpp#L6007-L6008) and ["getdata" message parsing](https://github.com/zcash/zcash/blob/8c95ceba46aa204f923f6c7d69a6bab912db9758/src/main.cpp#L6084-L6085) assume that the serialization of [CInv](https://github.com/zcash/zcash/blob/49ffee3f20b972dc3aa75e422c67523251cf088b/src/protocol.h#L123) (representing an inv type and hash) is fixed-length.

Reply:
We decided in the Arborist call that this is fine as-is.